### PR TITLE
Dialog: Only continue the freshest animations

### PR DESCRIPTION
**Pull request by ChrisAntaki**
This PR fixes an issue with the Dialog, where sometimes its stale animations overwrote its freshest animation

### CODE REVIEW

**Review by mmmicedcoffee — COMMENTED** [review]
is there a way we could test the new animation cancellation behavior?

**Comment by ChrisAntaki** [comment]
> is there a way we could test the new animation cancellation behavior?

Yep, expansion and collapse animations now have tests!

File: src/components/dialog.js [hunk 1]
```diff
@@ -342,41 +348,51 @@ export class Dialog {
...
+              },
+              300,
+              'ease-out'
+            ).then(() => {
+              if (animationNumber === this.animationNumber_) {
+                setImportantStyles(this.getElement(), {
+                  'height': `${newHeight}px`,
+                  'transform': 'translateY(0)',
+                });
+              }
```
[the remark this answers is not in the document]
Comment (ChrisAntaki) [reply src/components/dialog.js]: That's a good suggestion. I went ahead and updated the new code to use early returns where possible. It looks better, and it's helpful to see the `return` line's covered by tests

File: src/components/dialog-test.js [hunk 2]
```diff
@@ -155,38 +155,94 @@ describes.realWin('Dialog', {}, (env) => {
...
+      immediate();
+      await dialog.open();
+      await dialog.openView(view);
+
+      const setStylePropertySpy = sandbox
+        .stub(dialog.getElement().style, 'setProperty')
+        .callThrough();
+
+      dialog.resizeView(view, newHeight1, ANIMATE);
+      await dialog.resizeView(view, newHeight2, ANIMATE);
```
Comment (mmmicedcoffee) [inline src/components/dialog-test.js:223]: should the two new tests also have an expect statement verifying that the height property is set as expected?

File: src/components/dialog.js [hunk 3]
```diff
@@ -342,12 +348,23 @@ export class Dialog {
     }
     const newHeight = this.getMaxAllowedHeight_(height);
 
+    // Uniquely identify this animation.
+    // This lets callbacks abandon stale animations.
+    const animationNumber = ++this.animationNumber_;
+    const isStale = () => {
```
Comment (mmmicedcoffee) [inline src/components/dialog.js:354]: I like the addition of the new helper!

**Review by mmmicedcoffee — APPROVED** [review]

File: src/components/dialog-test.js (hunk 2 above)
Comment (ChrisAntaki) [reply src/components/dialog-test.js:223]: For sure, added!

**Pull request merged**